pkg/gofr/subscriber.go

```diff
@@ -24,35 +25,49 @@ func newSubscriptionManager(c *container.Container) SubscriptionManager {
 	}
 }
 
-func (s *SubscriptionManager) startSubscriber(topic string, handler SubscribeFunc) {
+func (s *SubscriptionManager) startSubscriber(ctx context.Context, topic string, handler SubscribeFunc) {
 	// continuously subscribe in an infinite loop
 	for {
-		msg, err := s.container.GetSubscriber().Subscribe(context.Background(), topic)
-		if msg == nil {
-			continue
-		}
+		msg, err := s.container.GetSubscriber().Subscribe(ctx, topic)
 
-		if errors.Is(err, kafka.ErrConsumerGroupNotProvided) {
-			s.container.Logger.Errorf("cannot subscribe as consumer_id is not provided in configs")
+		select {
+		case <-ctx.Done():
+			s.container.Logger.Infof("shutting down subscriber for topic %s", topic)
 			return
-		} else if err != nil {
-			s.container.Logger.Errorf("error while reading from topic %v, err: %v", topic, err.Error())
-			continue
+		default:
+			s.handleSubscription(topic, handler, msg, err)
 		}
+	}
+}
 
-		ctx := newContext(nil, msg, s.container)
-		err = func(ctx *Context) error {
-			// TODO : Move panic recovery at central location which will manage for all the different cases.
-			defer panicRecovery(ctx.Logger)
-			return handler(ctx)
-		}(ctx)
+func (s *SubscriptionManager) handleSubscription(topic string, handler SubscribeFunc, msg *pubsub.Message, err error) {
+	if errors.Is(err, kafka.ErrConsumerGroupNotProvided) {
+		s.container.Logger.Errorf("cannot subscribe as consumer_id is not provided in configs")
+		return
+	} else if err != nil {
```

Comment:
```suggestion
	}

if err != nil {
```
The return in the previous if allows to keep the code simple

Reply:
I tried not to change most code unrelated to shutdown. But this PR is getting so big that I'm afraid it's not gonna be merge at all